### PR TITLE
Fix for Issue 44762

### LINKED ISSUE #44762: LoadVector intrinsics get misoptimized inside loops on .NET 5
### Description
Any vector load intrinsics seem to get misoptimized when reusing the memory load address inside a loop on .NET 5. The issue can be reproduced with the code below:
```CS
using System;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

namespace IntrinsicsMisoptimizationTest {
    class Program {
        unsafe static void WriteArray (float* ptr, int count) {
            Console.Write ("[");
            for (int i = 0; i < count; i++) {
                if (i > 0)
                    Console.Write (", ");

                Console.Write (ptr [i]);
            }
            Console.WriteLine ("]");
        }

        unsafe static void Main (string [] args) {
            const int AvxVecLen = 4 * 2;

            var mem = stackalloc float [4 * 2];
            var memSpan = new Span<float> (mem, AvxVecLen);
            for (int i = 0; i < 1; i++) {
                Vector256<float> x1, x2;
                if (Avx.IsSupported) {
                    memSpan.Fill (5);
                    x1 = Avx.LoadVector256 (mem);

                    memSpan.Fill (15);
                    x2 = Avx.LoadVector256 (mem);

                    var x3 = Avx.Add (x1, x2);

                    Avx.Store (mem, x3);
                    WriteArray (mem, AvxVecLen);
                }
            }

            Console.ReadKey ();
        }
    }
}
```
This does not happen with optimizations disabled. (be it for the entire project or only for the function using these intrinsics)
It happens with both `Avx.LoadVector256` and `Sse.LoadVector128`, so I believe this happens with all x86 SIMD load intrinsics.

### Configuration
Tested on .NET 5.0.100, on Windows 10 x64, with a 7th gen Core i7 x64 CPU (i7 7700K). Not tested with non-x86 intrinsics. Seems specific to optimized builds.

### Regression?
This works properly in .NET Core 3.1.

### Other information
Partial disassembly from VS of the minimal repro sample given above: (I've added some missing source code lines for readability)
```
// [...]
            for (int i = 0; i < 1; i++) {
00007FF9B05E64C0  xor         edi,edi

                    memSpan.Fill (5);
00007FF9B05E64C2  lea         rcx,[rsp+50h]
00007FF9B05E64C7  vmovss      xmm1,dword ptr [IntrinsicsMisoptimizationTest.Program.Main(System.String[])+0D8h (07FF9B05E6548h)]
00007FF9B05E64CF  call        CLRStub[MethodDescPrestub]@7ff9b05e5a58 (07FF9B05E5A58h)
                    x1 = Avx.LoadVector256 (mem);
00007FF9B05E64D4  vmovups     ymm1,ymmword ptr [rsi]

                    memSpan.Fill (15);
00007FF9B05E64D8  lea         rcx,[rsp+50h]
00007FF9B05E64DD  vmovss      xmm1,dword ptr [IntrinsicsMisoptimizationTest.Program.Main(System.String[])+0DCh (07FF9B05E654Ch)]
00007FF9B05E64E5  call        CLRStub[MethodDescPrestub]@7ff9b05e5a58 (07FF9B05E5A58h)
                    x2 = Avx.LoadVector256 (mem);
00007FF9B05E64EA  vmovups     ymm0,ymmword ptr [rsi]

                    x3 = Avx.Add (x1, x2);
00007FF9B05E64EE  vaddps      ymm0,ymm0,ymm0

                    Avx.Store (mem, x3);
00007FF9B05E64F2  vmovups     ymmword ptr [rsi],ymm0
// [...]
```
As you can see, it incorrectly adds ymm0 to itself instead of adding ymm1 to ymm0.

Additionally, here's a section of the disassembly from my actual code, where I found this, as the issue manifests itself differently there:
```
// Normal
EchelonUtilities.SIMDLoadVectorIntoSpan (normal, memHigh);
    vmovss      dword ptr [rcx],xmm7
    vmovss      dword ptr [rcx+4],xmm8
    vmovss      dword ptr [rcx+8],xmm9
    vxorps      xmm11,xmm11,xmm11
    vmovss      dword ptr [rcx+0Ch],xmm11
EchelonUtilities.SIMDLoadVectorIntoSpan (normal, memLow);
    /* [...] */
normalReg = Avx.LoadVector256 (mem1);
    vmovups     ymm7,ymmword ptr [r14]
// D
EchelonUtilities.SIMDLoadScalarIntoSpan (d, memHigh);
    vmovss      dword ptr [rcx],xmm10
    vmovss      dword ptr [rcx+4],xmm10
    vmovss      dword ptr [rcx+8],xmm10
    vmovss      dword ptr [rcx+0Ch],xmm10
EchelonUtilities.SIMDLoadScalarIntoSpan (d, memLow);
    /* [...] */
dReg = Avx.LoadVector256 (mem1);
    vmovups     ymm7,ymmword ptr [r14]
```
As you can see, here it outright loads the data into the same ymm register.

### PR DESCRIPTION
**Pull request by briansull**
Treat both SIMD and HWIntrinsic Memory Loads as Byref Exposed Loads that depend upon the global byref memory state.
  In liveness added byref use for SIMD Memory Loads
  Corrected the comment for GenTreeHWIntrinsic::OperIsMemoryLoadOrStore()
  Added test case Runtime_44762.cs

### CODE REVIEW

**Comment by briansull** [comment]
@dotnet/jit-contrib PTAL

There were no Asm diffs in System.Private.Corelib 
and only two methods with minor diffs on Arm64 in the frameworks

**Comment by ViktorHofer** [comment]
// Auto-generated message

https://github.com/dotnet/runtime/commit/69e114c1abf91241a0eeecf1ecceab4711b8aa62 which was merged 12/7 removed the intermediate src/coreclr/**src**/ folder. This PR needs to be updated as it touches files in that directory which causes conflicts.

To update your commits you can use this bash script: https://gist.github.com/ViktorHofer/6d24f62abdcddb518b4966ead5ef3783. Feel free to use the comment section of the gist to improve the script for others.

**Comment by briansull** [comment]
@dotnet/jit-contrib PTAL  @AndyAyersMS 

Updated for src/coreclr/src
There were no Asm diffs in System.Private.Corelib or the full frameworks
The test case is passing

Fixes #44762
fgValueNumberHWIntrinsic is updated to make load intrinsic VNs dependent on the current heap VN

**Comment by briansull** [comment]
@tannergooding FYI

File: src/coreclr/jit/liveness.cpp [hunk 1]
```diff
@@ -300,6 +300,19 @@ void Compiler::fgPerNodeLocalVarLiveness(GenTree* tree)
             fgCurMemoryDef |= memoryKindSet(GcHeap, ByrefExposed);
             break;
 
+#ifdef FEATURE_SIMD
+        case GT_SIMD:
+        {
+            GenTreeSIMD* simdNode = tree->AsSIMD();
+            if (simdNode->OperIsMemoryLoad())
```
Comment (tannergooding) [inline src/coreclr/jit/liveness.cpp:307]: Should this also handle stores to ensure it matches the `case GT_HWINTRINSIC` path below?

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -8444,14 +8444,16 @@ void Compiler::fgValueNumberSimd(GenTree* tree)
...
+    VNFunc       simdFunc = GetVNFuncForNode(tree);
     ValueNumPair excSetPair;
     ValueNumPair normalPair;
 
     // There are some SIMD operations that have zero args, i.e.  NI_Vector128_Zero
     if (tree->AsOp()->gtOp1 == nullptr)
     {
         excSetPair = ValueNumStore::VNPForEmptyExcSet();
-        normalPair = vnStore->VNPairForFunc(tree->TypeGet(), GetVNFuncForNode(tree));
+        normalPair = vnStore->VNPairForFunc(tree->TypeGet(), simdFunc);
```
Comment (tannergooding) [inline src/coreclr/jit/valuenum.cpp:8456]: Do we need to do anything special for operations like `StoreFence`/`LoadFence`/`MemoryFence` which, while not load/stores likely shouldn't be CSE'd?

Comment (tannergooding) [reply src/coreclr/jit/valuenum.cpp:8456]: Ah, this is handling `GT_SIMD` not `GT_HWINTRINSIC` so the comment would apply elsewhere.

File: src/coreclr/jit/valuenum.cpp [hunk 3]
```diff
@@ -8585,9 +8596,37 @@ void Compiler::fgValueNumberHWIntrinsic(GenTree* tree)
...
-    int    lookupNumArgs    = HWIntrinsicInfo::lookupNumArgs(hwIntrinsicNode->gtHWIntrinsicId);
-    bool   encodeResultType = vnEncodesResultTypeForHWIntrinsic(hwIntrinsicNode->gtHWIntrinsicId);
-    VNFunc func             = GetVNFuncForNode(tree);
+    VNFunc func         = GetVNFuncForNode(tree);
+    bool   isMemoryLoad = hwIntrinsicNode->OperIsMemoryLoad();
+
+    // If we have a MemoryLoad operation we will use the fgValueNumberByrefExposedLoad
+    // method to assign a value number that depends upon fgCurMemoryVN[ByrefExposed] ValueNumber
+    //
+    if (isMemoryLoad)
```
Comment (tannergooding) [inline src/coreclr/jit/valuenum.cpp:8605]: This is for loads which have up to 2 arguments on x86/x64 but up to 3 arguments on ARM64: https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/hwintrinsiclistarm64.h#L171

Is any special handling needed for the GT_LIST? Likewise would the same be true for memory stores which can have 3+ on both x86/x64 and ARM64?

File: src/coreclr/jit/liveness.cpp (hunk 1 above)
Comment (tannergooding) [reply src/coreclr/jit/liveness.cpp:307]: Ah, we don't have `::OperIsMemoryStore` for `GT_SIMD` 👍

**Review by AndyAyersMS — COMMENTED** [review]
Does modelling HW intrinsics as always accessing byref exposed memeory handle the cases where we have HW intrinsic and non-HW intrinsic loads/stores to the same local intermixed (with aliasing via Unsafe.As or similar)? I wonder because we try and disambiguate the latter.

File: src/coreclr/jit/valuenum.cpp (hunk 2 above)
Comment (briansull) [reply src/coreclr/jit/valuenum.cpp:8456]: As this change might need to be backported to 5.0, I am limiting it to the minimal to fix the issue.

Currently all HWIntrinsic Loads are not CSE candidates,  I have an additional change that allows them to be CSE'd.
Only the non special ones can be made CSE candidates.

The special operations like PrefetchX, LoadFence,  etc.  will never be CSE candidates

(See OptCSE.cpp line 3677)

File: src/coreclr/jit/liveness.cpp (hunk 1 above)
Comment (briansull) [reply src/coreclr/jit/liveness.cpp:307]: Yeah, apparently there aren't any SIMD store instructions (and only one SIMD load instruction)

File: src/coreclr/jit/valuenum.cpp [hunk 4]
```diff
@@ -8585,9 +8596,37 @@ void Compiler::fgValueNumberHWIntrinsic(GenTree* tree)
         fgMutateGcHeap(tree DEBUGARG("HWIntrinsic - MemoryStore"));
```
Comment (briansull) [inline src/coreclr/jit/valuenum.cpp:8596]: @AndyAyersMS   All HW_INTRINSIC Memory Stores mutate the global heap and there are no SIMD store instructions

File: src/coreclr/jit/valuenum.cpp (hunk 3 above)
Comment (briansull) [reply src/coreclr/jit/valuenum.cpp:8605]: I will move up the check for GT_LIST to come before this.
We pessimize the ValueNumber when we have a  GT_LIST case.

**Comment by briansull** [comment]
@AndyAyersMS @tannergooding 
I updated the PR with the feedback from tTanner
Any more reviewer comments on this?

**Comment by briansull** [comment]
@dotnet/jit-contrib Someone - Please review and approve

**Review by AndyAyersMS — APPROVED** [review]
Thanks. This looks good.

**Pull request merged**